src/linux/cgroups.cpp

```diff
@@ -1425,16 +1424,15 @@ class TasksKiller : public Process<TasksKiller>
       const PID<TasksKiller>& pid)
   {
     // Cancel the freeze operation.
-    // TODO(jieyu): Wait until 'future' is in DISCARDED state before
-    // starting retry.
     future.discard();
 
-    // We attempt to kill the processes before we thaw again,
-    // due to a bug in the kernel. See MESOS-1758 for more details.
-    // We thaw the cgroup before trying to freeze again to allow any
-    // pending signals to be delivered. See MESOS-1689 for details.
-    // This is a short term hack until we have PID namespace support.
-    return Future<bool>(true)
+    // Wait until the freeze is cancelled, and then attempt to kill the
+    // processes before we thaw again, due to a bug in the kernel. See
+    // MESOS-1758 for more details.  We thaw the cgroup before trying to freeze
+    // again to allow any pending signals to be delivered. See MESOS-1689 for
+    // details.  This is a short term hack until we have PID namespace support.
+    return future
+      .recover([](const Future<Nothing>&){ return Future<Nothing>(Nothing()); })
```

Comment:
I think you intention is to wait until `future` is in `DISCARDED` state before starting retry, right? But if that happens, will we just return from here without calling the 3 methods below (i.e. no retry)?

Reply:
Hm, no?
We won't return, we carry on with the 3 methods below. I confirmed it by looking at logs.
The `[](const Future<Nothing>&){ return Future<Nothing>(Nothing()); }` is basically just a no-op in order to be able to call `recover`.
Is there a simpler way?

Reply:
> We won't return, we carry on with the 3 methods below.

I am not sure about it. For example, if [these code](https://github.com/apache/mesos/blob/1.11.0/3rdparty/libprocess/src/http.cpp#L1883:L1891) get called, will the `then` below it still be called?

Reply:
Yes, I would assume so.

Reply:
See this example:
```
cf@thinkpad:~/src/test/mesos$ cat test_process.cpp 
#include <process/process.hpp>

using namespace process;

int
main(int argc, char* argv[])
{
  auto promise = Promise<Nothing>();
  auto future = promise.future();

  future
    .recover([](const Future<Nothing>& future) -> Future<Nothing> {
      std::cout << "recover" << std::endl;
      return Nothing();
    })
    .then([](const Future<Nothing>& future) {
      std::cout << "then" << std::endl;
      return Nothing();
    });

  promise.discard();
  future.await();
}
cf@thinkpad:~/src/test/mesos$ /tmp/test_process 
recover
then
cf@thinkpad:~/src/test/mesos$ 
```